connection.go

```diff
@@ -0,0 +1,471 @@
+func (ct *ConnectionTransportTLS) Close() {
+	ct.mutex.Lock()
+	defer ct.mutex.Unlock()
+	if ct.conn != nil {
+		ct.conn.Close()
+	}
+}
+
+type LogTagsFromContext func(ctx context.Context) (map[interface{}]string, bool)
+
+// Connection encapsulates all client connection handling.
+type Connection struct {
+	srvAddr          string
+	handler          ConnectionHandler
+	transport        ConnectionTransport
+	errorUnwrapper   ErrorUnwrapper
+	reconnectBackoff *backoff.ExponentialBackOff
+	doCommandBackoff *backoff.ExponentialBackOff
+	wef              WrapErrorFunc
+	log              LogOutput
+	logPrefix        string
+	tagsFunc         LogTagsFromContext
+
+	// protects everything below.
+	mutex             sync.Mutex
+	client            GenericClient
+	server            *Server
+	reconnectChan     chan struct{}
+	reconnectErrPtr   *error             // Filled in with fatal reconnect err (if any) before reconnectChan is closed
+	cancelFunc        context.CancelFunc // used to cancel the reconnect loop
+	reconnectedBefore bool
+}
+
+// NewTLSConnection returns a connection that tries to connect to the
+// given server address with TLS.
+func NewTLSConnection(srvAddr string, rootCerts []byte,
+	errorUnwrapper ErrorUnwrapper, handler ConnectionHandler,
+	connectNow bool, l LogFactory, wef WrapErrorFunc, log LogOutput,
+	tagsFunc LogTagsFromContext) *Connection {
+	transport := &ConnectionTransportTLS{
+		rootCerts:  rootCerts,
+		srvAddr:    srvAddr,
+		logFactory: l,
+		wef:        wef,
+	}
+	return NewConnectionWithTransport(handler, transport, errorUnwrapper,
+		connectNow, wef, log, tagsFunc)
+}
+
+// NewConnectionWithTransport allows for connections with a custom
+// transport.
+func NewConnectionWithTransport(handler ConnectionHandler,
+	transport ConnectionTransport, errorUnwrapper ErrorUnwrapper,
+	connectNow bool, wef WrapErrorFunc, log LogOutput,
+	tagsFunc LogTagsFromContext) *Connection {
+	// retry w/exponential backoff
+	reconnectBackoff := backoff.NewExponentialBackOff()
+	// never give up reconnecting
+	reconnectBackoff.MaxElapsedTime = 0
+	randBytes := make([]byte, 4)
+	rand.Read(randBytes)
+	connectionPrefix := fmt.Sprintf("CONN %s %x", handler.HandlerName(),
+		randBytes)
+	connection := &Connection{
+		handler:          handler,
+		transport:        transport,
+		errorUnwrapper:   errorUnwrapper,
+		reconnectBackoff: reconnectBackoff,
+		doCommandBackoff: backoff.NewExponentialBackOff(),
+		log:              log,
+		logPrefix:        connectionPrefix,
+		tagsFunc:         tagsFunc,
+	}
+	if connectNow {
+		// start connecting now
+		connection.getReconnectChan()
+	}
+	return connection
+}
+
+// connect performs the actual connect() and rpc setup.
+func (c *Connection) connect(ctx context.Context) error {
+	c.log.Debug("%s: Connection: dialing transport", c.logPrefix)
+
+	// connect
+	transport, err := c.transport.Dial(ctx)
+	if err != nil {
+		c.log.Warning("%s: Connection: error dialing transport: %v",
+			c.logPrefix, err)
```

Comment:
I know it's pretty lame to manually prefix each log message.  I guess I could duplicate some of the logger logic to construct an internal `LogOutput` wrapper type that does the prefix automatically, but it doesn't seem worth it for just the few log lines in this file.  Thoughts?